File: Android/app/src/main/java/com/gokuencinar/iruniversal/flipper/FlipperIrCodec.kt
package com.gokuencinar.iruniversal.flipper

import com.gokuencinar.iruniversal.ir.IrCode
import com.gokuencinar.iruniversal.ir.IrProtocolEncoder
import java.util.UUID

data class ImportedIrSignal(
    val name: String,
    val code: IrCode,
    val sourceDescription: String
)

object FlipperIrCodec {
    fun parse(text: String): List<ImportedIrSignal> {
        val output = mutableListOf<ImportedIrSignal>()
        parseRecords(text).forEach { record ->
            val name = record["name"] ?: return@forEach
            val type = record["type"]?.lowercase() ?: return@forEach
            val id = "imported:" + UUID.randomUUID()

            if (type == "raw") {
                val frequency = record["frequency"]?.toIntOrNull() ?: return@forEach
                val durations = (record["data"] ?: "")
                    .split(Regex("\\s+"))
                    .mapNotNull { it.toIntOrNull() }
                    .filter { it > 0 }
                if (durations.size >= 2) {
                    output += ImportedIrSignal(
                        name,
                        IrCode(id, frequency, durations),
                        "Flipper RAW"
                    )
                }
            } else if (type == "parsed") {
                val protocol = record["protocol"] ?: return@forEach
                val address = parseHexBytes(record["address"].orEmpty())
                val command = parseHexBytes(record["command"].orEmpty())
                val code = IrProtocolEncoder.encode(protocol, address, command, id)
                if (code != null) {
                    output += ImportedIrSignal(name, code, "Flipper " + protocol)
                }
            }
        }
        return output
    }

    fun exportRaw(name: String, code: IrCode): String =
        exportRawRecords(listOf(name to code))

    fun exportRawRecords(records: List<Pair<String, IrCode>>): String {
        val lines = mutableListOf(
            "Filetype: IR signals file",
            "Version: 1"
        )
        records.forEach { (name, code) ->
            val safe = name.replace("\n", " ").replace("\r", " ")
            lines += "#"
            lines += "name: " + safe
            lines += "type: raw"
            lines += "frequency: " + code.effectiveCarrierHz
            lines += "duty_cycle: 0.330000"
            lines += "data: " + code.durationsMicros.joinToString(" ")
        }
        lines += ""
        return lines.joinToString("\n")
    }

    private fun parseRecords(text: String): List<Map<String, String>> {
        val records = mutableListOf<Map<String, String>>()
        var current = linkedMapOf<String, String>()
        var lastKey: String? = null

        fun finish() {
            if (current["name"] != null && current["type"] != null) {
                records += current.toMap()
            }
            current = linkedMapOf()
            lastKey = null
        }

        text.lineSequence().forEach { raw ->
            val line = raw.trim()
            if (line.isEmpty()) return@forEach
            if (line.startsWith("#")) {
                finish()
                return@forEach
            }

            val colon = line.indexOf(':')
            if (colon >= 0) {
                val key = line.substring(0, colon).trim().lowercase()
                val value = line.substring(colon + 1).trim()
                if (key == "name" && current["name"] != null) finish()
                current[key] = value
                lastKey = key
            } else if (lastKey == "data" && line.all { it.isDigit() || it.isWhitespace() }) {
                current["data"] = current["data"].orEmpty() + " " + line
            }
        }
        finish()
        return records
    }

    private fun parseHexBytes(value: String): List<Int> =
        value.split(Regex("\\s+"))
            .mapNotNull { token -> token.toIntOrNull(16)?.and(0xFF) }
}

object IrdbCsvCodec {
    fun parse(text: String, idPrefix: String = "irdb"): List<ImportedIrSignal> {
        val lines = text.lineSequence().toList()
        if (lines.size <= 1) return emptyList()

        return lines.drop(1).mapIndexedNotNull { index, line ->
            val fields = csv(line)
            if (fields.size < 5) return@mapIndexedNotNull null
            val device = fields[2].trim().toIntOrNull() ?: return@mapIndexedNotNull null
            val sub = fields[3].trim().toIntOrNull() ?: return@mapIndexedNotNull null
            val function = fields[4].trim().toIntOrNull() ?: return@mapIndexedNotNull null
            if (device !in 0..255 || function !in 0..255) return@mapIndexedNotNull null

            val proto = fields[1].uppercase().replace("-", "")
            val id = idPrefix + ":" + index + ":" + UUID.randomUUID()
            val code = when {
                proto.startsWith("NEC") && sub in 0..255 ->
                    IrProtocolEncoder.encode("NECEXT", listOf(device, sub), listOf(function, function.inv() and 0xFF), id)
                proto.startsWith("NEC") ->
                    IrProtocolEncoder.encode("NEC", listOf(device), listOf(function), id)
                proto.startsWith("RC5") ->
                    IrProtocolEncoder.encode("RC5", listOf(device), listOf(function), id)
                proto.startsWith("RC6") ->
                    IrProtocolEncoder.encode("RC6", listOf(device), listOf(function), id)
                proto.contains("SONY") || proto.contains("SIRC") ->
                    IrProtocolEncoder.encode(
                        if (proto.contains("20")) "SIRC20" else if (proto.contains("15")) "SIRC15" else "SIRC",
                        listOf(device), listOf(function), id
                    )
                proto == "JVC" || proto == "RCA" ->
                    IrProtocolEncoder.encode(proto, listOf(device), listOf(function), id)
                else -> null
            } ?: return@mapIndexedNotNull null

            val name = fields[0].trim().ifBlank { "BUTTON_" + function }
            ImportedIrSignal(name, code, "IRDB " + proto)
        }
    }

    private fun csv(line: String): List<String> {
        val out = mutableListOf<String>()
        val current = StringBuilder()
        var quoted = false
        var i = 0
        while (i < line.length) {
            when (val c = line[i]) {
                '"' -> {
                    if (quoted && i + 1 < line.length && line[i + 1] == '"') {
                        current.append('"')
                        i++
                    } else quoted = !quoted
                }
                ',' -> if (quoted) current.append(c) else {
                    out += current.toString()
                    current.clear()
                }
                else -> current.append(c)
            }
            i++
        }
        out += current.toString()
        return out
    }
}

File: Sources/FlipperIRCodec.swift
import Foundation
import SwiftUI
import UniformTypeIdentifiers

extension UTType {
    static let flipperIR =
        UTType(filenameExtension: "ir")
        ?? .plainText
}

struct ImportedIRSignal: Identifiable {
    let id = UUID()
    let name: String
    let code: IRCode
    let sourceDescription: String
}

enum FlipperIRCodec {
    static func parse(
        text: String
    ) -> [ImportedIRSignal] {
        let records =
            parseRecords(text)

        var output: [ImportedIRSignal] = []

        for record in records {
            guard
                let name =
                    record["name"],
                let type =
                    record["type"]?
                    .lowercased()
            else {
                continue
            }

            let id =
                "imported:\(UUID().uuidString)"

            if type == "raw" {
                guard
                    let frequencyText =
                        record["frequency"],
                    let frequency =
                        Int(frequencyText)
                else {
                    continue
                }

                let durations =
                    (record["data"] ?? "")
                    .split {
                        $0 == " "
                        || $0 == "\t"
                        || $0 == "\n"
                    }
                    .compactMap {
                        UInt32($0)
                    }

                guard durations.count >= 2 else {
                    continue
                }

                output.append(
                    ImportedIRSignal(
                        name: name,
                        code: IRCode(
                            id: id,
                            carrierHz: frequency,
                            durationsMicros:
                                durations
                        ),
                        sourceDescription:
                            "Flipper RAW"
                    )
                )

                continue
            }

            if type == "parsed" {
                let protocolName =
                    record["protocol"] ?? ""

                let address =
                    parseHexBytes(
                        record["address"] ?? ""
                    )

                let command =
                    parseHexBytes(
                        record["command"] ?? ""
                    )

                if let code =
                    IRProtocolEncoder.encode(
                        protocolName:
                            protocolName,
                        address: address,
                        command: command,
                        id: id
                    )
                {
                    output.append(
                        ImportedIRSignal(
                            name: name,
                            code: code,
                            sourceDescription:
                                "Flipper \(protocolName)"
                        )
                    )
                }
            }
        }

        return output
    }

    static func exportRaw(
        name: String,
        code: IRCode
    ) -> String {
        exportRawRecords(
            [(name, code)]
        )
    }

    static func exportRemote(
        _ remote: CustomRemote
    ) -> String {
        exportRawRecords(
            remote.buttons.map {
                ($0.name, $0.code)
            }
        )
    }

    private static func exportRawRecords(
        _ records: [(String, IRCode)]
    ) -> String {
        var lines = [
            "Filetype: IR signals file",
            "Version: 1",
        ]

        for (name, code) in records {
            let safeName =
                name
                .replacingOccurrences(
                    of: "\n",
                    with: " "
                )
                .replacingOccurrences(
                    of: "\r",
                    with: " "
                )

            let data =
                code.durationsMicros
                .map(String.init)
                .joined(separator: " ")

            lines.append("#")
            lines.append("name: \(safeName)")
            lines.append("type: raw")
            lines.append(
                "frequency: \(code.carrierHz == 0 ? 38000 : code.carrierHz)"
            )
            lines.append("duty_cycle: 0.330000")
            lines.append("data: \(data)")
        }

        lines.append("")
        return lines.joined(separator: "\n")
    }

    private static func parseRecords(
        _ text: String
    ) -> [[String: String]] {
        var records:
            [[String: String]] = []

        var current:
            [String: String] = [:]

        var lastKey: String?

        func finish() {
            if
                current["name"] != nil,
                current["type"] != nil
            {
                records.append(current)
            }

            current = [:]
            lastKey = nil
        }

        for rawLine
            in text.components(
                separatedBy: .newlines
            )
        {
            let line =
                rawLine.trimmingCharacters(
                    in: .whitespacesAndNewlines
                )

            if line.isEmpty {
                continue
            }

            if line.hasPrefix("#") {
                finish()
                continue
            }

            if let colon =
                line.firstIndex(of: ":")
            {
                let key =
                    String(
                        line[..<colon]
                    )
                    .trimmingCharacters(
                        in: .whitespacesAndNewlines
                    )
                    .lowercased()

                let value =
                    String(
                        line[
                            line.index(
                                after: colon
                            )...
                        ]
                    )
                    .trimmingCharacters(
                        in: .whitespacesAndNewlines
                    )

                if
                    key == "name",
                    current["name"] != nil
                {
                    finish()
                }

                current[key] = value
                lastKey = key
                continue
            }

            if
                lastKey == "data",
                line.allSatisfy({
                    $0.isNumber
                    || $0 == " "
                    || $0 == "\t"
                })
            {
                current["data"] =
                    (
                        current["data"]
                        ?? ""
                    )
                    + " "
                    + line
            }
        }

        finish()

        return records
    }

    private static func parseHexBytes(
        _ value: String
    ) -> [UInt8] {
        value
            .split {
                $0 == " "
                || $0 == "\t"
            }
            .compactMap {
                UInt8(
                    $0,
                    radix: 16
                )
            }
    }
}

struct FlipperIRDocument: FileDocument {
    static var readableContentTypes:
        [UTType] {
        [
            .flipperIR,
            .plainText,
            .data,
        ]
    }

    var text: String

    init(
        text: String = ""
    ) {
        self.text = text
    }

    init(
        configuration:
            ReadConfiguration
    ) throws {
        if
            let data =
                configuration.file
                .regularFileContents,
            let value =
                String(
                    data: data,
                    encoding: .utf8
                )
        {
            text = value
        } else {
            text = ""
        }
    }

    func fileWrapper(
        configuration:
            WriteConfiguration
    ) throws -> FileWrapper {
        FileWrapper(
            regularFileWithContents:
                text.data(
                    using: .utf8
                ) ?? Data()
        )
    }
}
